### DISCUSSION TITLE
ignorePaths in packageRules seems to be ignored

### DISCUSSION THREAD

**Discussion by y-nk** [renovatebot/renovate#33007] [Request Help] [answered]
### How are you running Renovate?

A Mend.io-hosted app

### Please tell us more about your question or problem

Hello,

In our monorepo, I've set a renovatebot config to have a few separate jobs:

1. one for our unique rust service
2. one for our node packages
3. one for our node services

i've set each to a separate `packageRules` entry with explicit `includePaths` for the 2 first jobs and one complementary `ignorePaths` for the last job... but that one doesn't seem to work.

Subset of the config here:

```json5
  "packageRules": [
    {
      // graphql gateway
      "managers": ["cargo", "nodenv", "npm"],
      "groupName": "graphql gateway update",
      "schedule": ["on the 1st and 3rd day instance on monday"], // every 2 weeks
      "includePaths": ["./services/graphql-gateway"],
    },
    {
      // our packages
      "managers": ["nodenv", "npm"],
      "groupName": "packages updates",
      "updateLockFiles": false,
      "schedule": ["on the 1st and 3rd day instance on monday"],
      "includePaths": ["./packages"]
    },
    {
      // the rest of node (services, jobs, utils)
      "managers": ["nodenv", "npm"],
      "groupName": "general update",
      "updateLockFiles": false,
      "schedule": ["on the 2nd and 4th day instance on monday"],
      "ignorePaths": [
        "**/services/graphql-gateway/**" // ignore 1st job
        "**/packages/**", // ignore 2nd job
      ],
    },
    {
      // nodenv are always pinned, so we want to keep updating them like this
      "matchManagers": ["nodenv"],
      "updatePinnedDependencies": true,
    }
  ],

```

(renovate logs below)

PR ends up like this:

<img width="743" alt="image" src="https://github.com/user-attachments/assets/9c7bb170-75eb-4c26-8c05-99ba5492c4b4">

Any idea what went wrong? I'm thinking to switch to `includePaths` but the spirit behind that last job is to gather all which were not taken by the 2 others.

### Logs (if relevant)

<details><summary>Logs (truncated)</summary>

```
...

DEBUG: Existing config file confirmed
DEBUG: Repository config
{
  "fileName": ".github/renovate.json5"
  "config": {
    "$schema": "https://docs.renovatebot.com/renovate-schema.json",
    "extends": [
      ":prConcurrentLimit10",
      "workarounds:typesNodeVersioning"
    ],
    "enabledManagers": [
      "cargo",
      "nodenv",
      "npm"
    ],
    "dependencyDashboard": true,
    "labels": [
      "no-fail-fast",
      "skip-audit"
    ],
    "npmrc": "***********",
    "timezone": "Asia/Bangkok",
    "major": {
      "enabled": false
    },
    "rangeStrategy": "bump",
    "updatePinnedDependencies": false,
    "hostRules": [
      {
        "matchHost": "https://npm.pkg.github.com/",
        "hostType": "npm",
        "encrypted": {
          "token": "***********"
        }
      }
    ],
    "packageRules": [
      {
        "managers": [
          "cargo",
          "nodenv",
          "npm"
        ],
        "groupName": "graphql gateway update",
        "schedule": [
          "on the 1st and 3rd day instance on monday"
        ],
        "includePaths": [
          "./services/graphql-gateway"
        ]
      },
      {
        "managers": [
          "nodenv",
          "npm"
        ],
        "groupName": "packages updates",
        "updateLockFiles": false,
        "schedule": [
          "on the 1st and 3rd day instance on monday"
        ],
        "includePaths": [
          "./packages"
        ]
      },
      {
        "managers": [
          "nodenv",
          "npm"
        ],
        "groupName": "general update",
        "updateLockFiles": false,
        "schedule": [
          "on the 2nd and 4th day instance on monday"
        ],
        "ignorePaths": [
          "**/packages/**",
          "**/services/graphql-gateway/**"
        ]
      },
      {
        "matchManagers": [
          "nodenv"
        ],
        "updatePinnedDependencies": true
      }
    ]
  }
}
...
DEBUG: Found repo ignorePaths
{
  "ignorePaths": [
    "**/node_modules/**",
    "**/bower_components/**"
  ]
}
...
DEBUG: Branch summary
{
  "cacheModified": true
  "baseBranches": [
    {
      "branchName": "master",
      "sha": "c97eb0282603a2c69883c2e8e416ee3a8b77914a"
    }
  ]
  "branches": [
    {
      "automerge": false,
      "baseBranch": "master",
      "baseBranchSha": "c97eb0282603a2c69883c2e8e416ee3a8b77914a",
      "branchName": "renovate/general-update",
      "branchSha": "40b5f771ddc9b7feeeaae14d6536902ee09c0911",
      "isModified": true,
      "isPristine": false
    }
  ]
  "defaultBranch": "master"
  "inactiveBranches": [
    "renovate/graphql-gateway-update"
  ]
}

DEBUG: branches info extended
{
  "branchesInformation": [
    {
      "branchName": "renovate/general-update",
      "prNo": 3007,
      "prTitle": "chore(deps): update general update",
      "result": "pr-edited",
      "upgrades": [
        {
          "datasource": "npm",
          "depName": "@grpc/grpc-js",
          "displayPending": "",
          "fixedVersion": "1.12.2",
          "currentVersion": "1.12.2",
          "currentValue": "^1.12.2",
          "newValue": "^1.12.4",
          "newVersion": "1.12.4",
          "packageFile": "packages/nestjs-starterkit/package.json",
          "updateType": "patch",
          "packageName": "@grpc/grpc-js"
        },
```

</details>

**Comment by rarkins** [comment]
`includePaths`/`ignorePaths` are not intended for use in `packageRules`. They are used as part of the `extract` phase of Renovate, before any `packageRules` are evaluated. If you wish them to be applied to certain managers only, then you can nest them inside the applicable object, e.g.

```json
{
  "npm": {
       "ignorePaths": [
        "**/services/graphql-gateway/**",
        "**/packages/**",
      ]
  }
}
```

**Reply by y-nk** [reply]
i understand. the goal was not to limit it to a specific manager but rather to a specific job. as i understand, there should be no way for this to be done? would that be an acceptable feature (for which we can eventually help/produce a PR)?

**Reply by rarkins** [reply]
Please describe your requirements in as plain english as possible, also avoiding Renovate terminology. i.e. tell me in non-renovate terms what you're "really trying to do". Maybe a solution already exists

**Reply by y-nk** [reply]
"in a monorepo, we're trying to have 3 separate PRs for updating dependencies selecting 3 separate globs, each of them containing one or multiple monorepo workspaces. each of those PR could include more than one runtime (node, rust)."

i can't think of anything clearer but happy to iterate on it

EDIT: found [this issue](https://github.com/renovatebot/renovate/discussions/15711), which led to [this changelog](https://arc.net/l/quote/gbbpfrrw), ultimately ending to [`matchFileNames`](https://docs.renovatebot.com/configuration-options/#matchfilenames). should it be the way?

**Reply by rarkins** [reply, marked as answer]
Yes, if you're using `packageRules` then `matchFileNames` is what should be used.

If you read the docs, you can see all supported matchers quite easily:

<img width="277" alt="image" src="https://github.com/user-attachments/assets/646165bb-9c5e-45d1-9346-3d0fe974c6fe">

**Reply by y-nk** [reply]
thank you. i have read the docs quite often but since some of the top level props are also available in `packageRules` i couldn't make sure which one would/wouldn't work and wrongly assumed `includePaths` and `ignorePaths` would be.

**Reply by rarkins** [reply]
I was more referring to being able to "discover" the matchX ones. Btw the matchJsonata one is new and very powerful although could do with some more examples

**Reply by y-nk** [reply]
it looks quite powerful indeed. i will try to look at the JSONata syntax. thanks a lot for this suggestion
